src/backends/caffe2/caffe2lib.cc

```diff
@@ -87,176 +65,416 @@ namespace dd {
+  void Caffe2Lib<TInputConnectorStrategy,TOutputConnectorStrategy,TMLModel>::
+  instantiate_template(const APIData &ad) {
+    this->_mlmodel.update_from_repository(this->_logger);
+
+    // Assert that there is no risk of erasing model files
+    if (!this->_mlmodel._init.empty() && !this->_mlmodel._predict.empty()) {
+      throw MLLibBadParamException("using template while model files already exists, "
+				   "remove 'template' from 'mllib', "
+				   "or remove existing '.pb' files ?");
+    }
+
+    //XXX Implement build-in template creation (mlp, convnet, resnet, etc.)
+
+    std::string model_tmpl = ad.get("template").get<std::string>();
+    std::string source = this->_mlmodel._mlmodel_template_repo + '/' + model_tmpl;
+    std::vector<std::string> files({"predict_net.pb", "init_net.pb"});
+    this->_logger->info("instantiating model template {}", model_tmpl);
+    this->_logger->info("source = {}", source);
+    this->_logger->info("dest = {}", this->_mlmodel._repo);
+
+    // Copy files
+    for (const std::string &file : files) {
+      std::string src = source + "/" + file;
+      std::string dst = this->_mlmodel._repo + "/" + file;
+      switch (fileops::copy_file(src, dst)) {
+      case 1: throw MLLibBadParamException("failed to locate model template " + src);
+      case 2: throw MLLibBadParamException("failed to create model template destination " + dst);
+      }
+    }
+
+    //XXX Configure the nets protobuf files
+
+    // Update the mlmodel
+    this->_mlmodel._model_template = model_tmpl;
+    this->_mlmodel.update_from_repository(this->_logger);
+  }
+
+  template <class TInputConnectorStrategy, class TOutputConnectorStrategy, class TMLModel>
+  void Caffe2Lib<TInputConnectorStrategy,TOutputConnectorStrategy,TMLModel>::
+  set_train_mode(const APIData &ad, bool train) {
+
+    // Update the input connector
+    TInputConnectorStrategy inputc(this->_inputc);
+    inputc._train = train;
     try {
-      ids = { ad.get("gpuid").get<int>() };
-    } catch(std::exception &e) {
-      ids = ad.get("gpuid").get<std::vector<int>>();
+      inputc.transform(ad);
+    } catch (...) {
+      this->_logger->info("Could not configure the InputConnector");
+      throw;
     }
-    if (ids.size() == 1 && ids[0] == -1) {
-      ids.clear();
-      int count_gpus = 0;
-      cudaGetDeviceCount(&count_gpus);
-      ids.resize(count_gpus);
-      std::iota(ids.begin(), ids.end(), 0);
+
+    // Update the local input connector
+    bool force_init = inputc.needs_reconfiguration(_last_inputc);
+    _last_inputc = inputc;
+
+    // Reset the state
+    _state.reset();
+    _state.set_is_training(train);
+    if (force_init) {
+      _state.force_init();
     }
-    CAFFE_ENFORCE(!ids.empty());
-    for (int id: ids) {
-      this->_logger->info("Using GPU {}", id);
+  }
+
+  //XXX When classifying, the last layer is currently not reshaped based on _nclasses
```

Comment:
The layer before the softmax needs to be reshaped based on `_nclasses`. Otherwise the model templates cannot be used for problems that expect a number of classes not exactly the same as the template.
I believe the reshape is mandatory, unfortunately.